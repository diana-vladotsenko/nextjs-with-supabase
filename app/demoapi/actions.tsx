"use server";

import { Cat } from "@/types/Cat";

export async function fetchCats() {
  const res = await fetch(
    "https://api.thecatapi.com/v1/images/search?limit=10",
    { cache: "no-store" }
  );
  if (!res.ok) throw new Error("Failed to fetch cats");
  const data: Cat[] = await res.json();
  return data;
}
