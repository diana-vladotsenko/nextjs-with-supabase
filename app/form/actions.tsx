"use server";

// import { revalidatePath } from "next/cache";
// import { createClient } from "@/lib/supabase/server";

export async function example() {}
