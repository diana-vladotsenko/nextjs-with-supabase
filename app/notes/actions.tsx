"use server";

import { revalidatePath } from "next/cache";
import { createClient } from "@/lib/supabase/server";

export async function createNote(formData: FormData) {
  const supabase = await createClient();
  const title = String(formData.get("title") || "").trim();
  if (!title) throw new Error("Title is required");

  const { error } = await supabase.from("notes").insert({ title });
  if (error) throw error;

  revalidatePath("/notes");
}

export async function deleteNote(formData: FormData) {
  const supabase = await createClient();
  const idStr = String(formData.get("id") || "");
  const id = Number(idStr);
  if (!Number.isFinite(id)) throw new Error("Invalid id");

  const { error } = await supabase.from("notes").delete().eq("id", id);
  if (error) throw error;

  revalidatePath("/notes");
}
