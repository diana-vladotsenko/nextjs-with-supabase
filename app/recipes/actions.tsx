"use server";

import { revalidatePath } from "next/cache";
import { createClient } from "@/lib/supabase/server";

export async function addRecipe(fd: FormData) {
  try {
    const title = String(fd.get("title") || "").trim();
    const ingredients = String(fd.get("ingredients") || "").trim();
    const instruction = String(fd.get("instruction") || "").trim();

    if (!title || !ingredients || !instruction) return;

    const supabase = await createClient();
    const { error } = await supabase.from("recipe").insert({
      title: title,
      ingredients: ingredients,
      instruction: instruction,
      created_at: new Date().toISOString(),
    });

    if (error) throw new Error(error.message);
    revalidatePath("/recipes");
  } catch (err) {
    throw err;
  }
}

export async function deleteRecipe(fd: FormData) {
  try {
    const id = Number(fd.get("id") ?? "");

    if (!id) return;

    const supabase = await createClient();
    const { error } = await supabase.from("recipe").delete().eq("id", id);

    if (error) throw new Error(error.message);

    revalidatePath("/recipes");
  } catch (err) {
    throw err;
  }
}

export async function updateRecipe(fd: FormData) {
  try {
    const id = Number(fd.get("id"));
    const title = String(fd.get("title") || "").trim();
    const ingredients = String(fd.get("ingredients") || "").trim();
    const instruction = String(fd.get("instruction") || "").trim();

    if (!id || !title || !ingredients || !instruction) return;

    const supabase = await createClient();
    const { error } = await supabase
      .from("recipe")
      .update({ title })
      .eq("id", id);

    if (error) throw new Error(error.message);

    revalidatePath("/recipes");
  } catch (err) {
    throw err;
  }
}
