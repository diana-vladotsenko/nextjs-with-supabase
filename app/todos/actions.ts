"use server";

import { revalidatePath } from "next/cache";
import { createClient } from "@/lib/supabase/server";

export async function addTask(formData: FormData) {
  try {
    const name = String(formData.get("name") ?? "").trim();
    
    if (!name) return;

    const supabase = await createClient();
    const { error } = await supabase.from("tasks").insert({ name: name , is_done: false});

    if (error) throw new Error(error.message);

    revalidatePath("/todos");
  } catch (err) {
    throw err;
  }
}

export async function completeTask(formData: FormData) {
  try {
    const id = Number(formData.get("id") ?? "");

    if (!id) return;

    const supabase = await createClient();
    const { error } = await supabase
      .from("tasks")
      .update({ is_done: true })
      .eq("id", id);

    if (error) throw new Error(error.message);
    revalidatePath("/todos");
  } catch (err) {
    throw err;
  }
}

export async function deleteTask(formData: FormData) {
  try {
    const id = Number(formData.get("id") ?? "");
    if (!id) return;
    
    const supabase = await createClient();
    const { error } = await supabase
      .from("tasks")
      .delete()
      .eq("id", id);

    if (error) throw new Error(error.message);

    revalidatePath("/todos");
  } catch (err) {
    throw err;
  }
}

export async function updateTask(formData: FormData) {
  const id = Number(formData.get("id"));
  const name = String(formData.get("name") ?? "").trim();

  if (!id || !name) return;

  const supabase = await createClient();
  const { error } = await supabase
    .from("tasks")
    .update({ name })
    .eq("id", id);

  if (error) throw new Error(error.message);

  revalidatePath("/todos");
}