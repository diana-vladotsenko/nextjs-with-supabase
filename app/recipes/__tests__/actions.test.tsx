/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, it, expect, vi, beforeEach } from "vitest";

const { store } = vi.hoisted(() => ({
  store: {
    op: "ok" as "ok" | "insert_error" | "update_error" | "delete_error",
    last: {
      table: "" as string,
      payload: null as any,
      eq: null as { column: string; value: any } | null,
    },
  },
}));

vi.mock("next/cache", () => {
  return {
    revalidatePath: vi.fn(),
  };
});

vi.mock("@/lib/supabase/server", () => {
  const fromSpy = vi.fn((table: string) => {
    store.last.table = table;

    const insert = vi.fn(async (payload: any) => {
      store.last.payload = payload;
      if (store.op === "insert_error") return { error: { message: "insert failed" } };
      return { error: null };
    });

    const update = vi.fn((_payload: any) => {
      store.last.payload = _payload;
      return {
        eq: vi.fn(async (column: string, value: any) => {
          store.last.eq = { column, value };
          if (store.op === "update_error") return { error: { message: "update failed" } };
          return { error: null };
        }),
      };
    });

    const _delete = vi.fn(() => {
      return {
        eq: vi.fn(async (column: string, value: any) => {
          store.last.eq = { column, value };
          if (store.op === "delete_error") return { error: { message: "delete failed" } };
          return { error: null };
        }),
      };
    });

    return { insert, update, delete: _delete };
  });

  const createClient = vi.fn(async () => ({
    from: fromSpy,
  }));

  return { createClient, __spies: { fromSpy } };
});

import { revalidatePath } from "next/cache";
import { addRecipe, deleteRecipe, updateRecipe } from "../actions";

describe("recipes/actions", () => {
  beforeEach(() => {
    store.op = "ok";
    store.last = { table: "", payload: null, eq: null };
    vi.clearAllMocks();
  });

  describe("addRecipe", () => {
    it("returns early when any required field is empty", async () => {
      const fd = new FormData();
      fd.set("title", "   ");
      fd.set("ingredients", "flour, water");
      fd.set("instruction", "mix and bake");
      await expect(addRecipe(fd)).resolves.toBeUndefined();
      expect((revalidatePath as any).mock.calls.length).toBe(0);
    });

    it("inserts title, ingredients, instruction and revalidates", async () => {
      const fd = new FormData();
      fd.set("title", "Bread");
      fd.set("ingredients", "flour, water, yeast, salt");
      fd.set("instruction", "mix, proof, bake");
      await addRecipe(fd);
      expect(store.last.table).toBe("recipe");
      expect(store.last.payload.title).toBe("Bread");
      expect(store.last.payload.ingredients).toBe("flour, water, yeast, salt");
      expect(store.last.payload.instruction).toBe("mix, proof, bake");
      expect(typeof store.last.payload.created_at).toBe("string");
      expect(revalidatePath).toHaveBeenCalledWith("/recipes");
    });

    it("propagates insert error", async () => {
      store.op = "insert_error";
      const fd = new FormData();
      fd.set("title", "X");
      fd.set("ingredients", "Y");
      fd.set("instruction", "Z");
      await expect(addRecipe(fd)).rejects.toThrow("insert failed");
    });
  });

  describe("deleteRecipe", () => {
    it("returns early when id is missing or 0", async () => {
      const fd = new FormData();
      await expect(deleteRecipe(fd)).resolves.toBeUndefined();
      expect((revalidatePath as any).mock.calls.length).toBe(0);
    });

    it("deletes row by id and revalidates", async () => {
      const fd = new FormData();
      fd.set("id", "9");
      await deleteRecipe(fd);
      expect(store.last.table).toBe("recipe");
      expect(store.last.eq).toEqual({ column: "id", value: 9 });
      expect(revalidatePath).toHaveBeenCalledWith("/recipes");
    });

    it("propagates delete error", async () => {
      store.op = "delete_error";
      const fd = new FormData();
      fd.set("id", "3");
      await expect(deleteRecipe(fd)).rejects.toThrow("delete failed");
    });
  });

  describe("updateRecipe", () => {
    it("returns early when id missing or any required field empty", async () => {
      const fd = new FormData();
      fd.set("id", "1");
      fd.set("title", "Bread");
      fd.set("ingredients", "   ");
      fd.set("instruction", "bake");
      await expect(updateRecipe(fd)).resolves.toBeUndefined();
      expect((revalidatePath as any).mock.calls.length).toBe(0);
    });

    it("updates title by id and revalidates", async () => {
      const fd = new FormData();
      fd.set("id", "12");
      fd.set("title", "Sourdough");
      fd.set("ingredients", "flour, water, starter, salt");
      fd.set("instruction", "mix, bulk, shape, proof, bake");
      await updateRecipe(fd);
      expect(store.last.table).toBe("recipe");
      expect(store.last.payload).toEqual({ title: "Sourdough" });
      expect(store.last.eq).toEqual({ column: "id", value: 12 });
      expect(revalidatePath).toHaveBeenCalledWith("/recipes");
    });

    it("propagates update error", async () => {
      store.op = "update_error";
      const fd = new FormData();
      fd.set("id", "5");
      fd.set("title", "X");
      fd.set("ingredients", "Y");
      fd.set("instruction", "Z");
      await expect(updateRecipe(fd)).rejects.toThrow("update failed");
    });
  });
});
