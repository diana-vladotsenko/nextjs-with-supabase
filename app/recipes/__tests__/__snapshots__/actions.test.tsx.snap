// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loading skeleton > matches snapshot 1`] = `
<div
  class="max-w-3xl mx-auto p-6 space-y-4 animate-pulse"
>
  <div
    class="h-6 w-40 rounded bg-neutral-200"
  />
  <div
    class="space-y-2"
  >
    <div
      class="h-10 w-full rounded bg-neutral-200"
    />
    <div
      class="h-10 w-full rounded bg-neutral-200"
    />
    <div
      class="h-20 w-full rounded bg-neutral-200"
    />
    <div
      class="h-8 w-24 rounded bg-neutral-200"
    />
  </div>
  <div
    class="space-y-3 pt-6"
  >
    <div
      class="h-24 rounded-xl bg-neutral-200"
    />
    <div
      class="h-24 rounded-xl bg-neutral-200"
    />
  </div>
</div>
`;
